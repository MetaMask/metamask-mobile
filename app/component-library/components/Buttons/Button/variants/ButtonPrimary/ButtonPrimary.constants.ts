/* eslint-disable import/prefer-default-export */

// External dependencies.
import { SAMPLE_BUTTONBASE_PROPS } from '../../foundation/ButtonBase/ButtonBase.constants';

// Internal dependencies.
import { ButtonPrimaryProps } from './ButtonPrimary.types';

// Samples
export const SAMPLE_BUTTONPRIMARY_PROPS: ButtonPrimaryProps = {
  ...SAMPLE_BUTTONBASE_PROPS,
};
