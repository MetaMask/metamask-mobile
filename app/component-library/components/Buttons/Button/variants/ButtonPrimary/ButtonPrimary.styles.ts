// Third party dependencies.
import { StyleSheet, ViewStyle } from 'react-native';

// External dependencies.
import { Theme } from '../../../../../../util/theme/models';

/**
 * Style sheet function for ButtonPrimary component.
 *
 * @param params Style sheet params.
 * @param params.vars Inputs that the style sheet depends on.
 * @returns StyleSheet object.
 */
const styleSheet = (params: { theme: Theme; vars: any }) => {
  const { vars, theme } = params;
  const { colors } = theme;
  const { style, isDanger, pressed } = vars;
  const backgroundColorObj = isDanger ? colors.error : colors.primary;
  const backgroundColor: string = pressed
    ? backgroundColorObj.alternative
    : backgroundColorObj.default;

  return StyleSheet.create({
    base: Object.assign(
      {
        backgroundColor,
      } as ViewStyle,
      style,
    ) as ViewStyle,
  });
};

export default styleSheet;
