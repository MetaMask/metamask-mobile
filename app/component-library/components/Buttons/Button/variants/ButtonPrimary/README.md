# ButtonPrimary

ButtonPrimary is used for primary call to actions.

## Props

This component extends [ButtonBaseProps](../ButtonBase/ButtonBase.types.ts#L14) from [ButtonBase](../ButtonBase/ButtonBase.tsx) component.

## Usage

```javascript
<ButtonPrimary
  label={SAMPLE_LABEL}
  iconName={IconName.BankFilled}
  size={ButtonSize.Md}
  onPress={ONPRESS_HANDLER}
  isDanger
  width={ButtonWidthTypes.Full}
/>;
```
