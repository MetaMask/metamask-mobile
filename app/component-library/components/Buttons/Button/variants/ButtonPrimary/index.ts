export { default } from './ButtonPrimary';
export { ButtonPrimaryVariants } from './ButtonPrimary.types';
