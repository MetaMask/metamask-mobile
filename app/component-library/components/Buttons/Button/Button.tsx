/* eslint-disable react/prop-types */
import React from 'react';

// External dependencies.
import ButtonLink from './variants/ButtonLink';
import ButtonPrimary from './variants/ButtonPrimary';
import ButtonSecondary from './variants/ButtonSecondary';

// Internal dependencies.
import { ButtonProps, ButtonVariants } from './Button.types';

const Button = (buttonProps: ButtonProps) => {
  switch (buttonProps.variant) {
    case ButtonVariants.Link:
      return <ButtonLink {...buttonProps} />;
    case ButtonVariants.Primary:
      return <ButtonPrimary {...buttonProps} />;
    case ButtonVariants.Secondary:
      return <ButtonSecondary {...buttonProps} />;
    default:
      throw new Error('Invalid Button Variant');
  }
};

export default Button;
