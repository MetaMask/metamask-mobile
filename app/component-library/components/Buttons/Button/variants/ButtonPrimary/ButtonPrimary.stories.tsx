/* eslint-disable no-console */

// Third party dependencies.
import React from 'react';
import { select, text, boolean } from '@storybook/addon-knobs';

// External dependencies.
import { storybookPropsGroupID } from '../../../../../constants/storybook.constants';
import { IconName } from '../../../../Icon';
import { ButtonSize, ButtonWidthTypes } from '../../Button.types';

// Internal dependencies.
import ButtonPrimary from './ButtonPrimary';
import { ButtonPrimaryProps } from './ButtonPrimary.types';

export const getButtonPrimaryStoryProps = (): ButtonPrimaryProps => {
  const sizeSelector = select(
    'size',
    ButtonSize,
    ButtonSize.Md,
    storybookPropsGroupID,
  );
  const widthSelector = select(
    'width',
    ButtonWidthTypes,
    ButtonWidthTypes.Auto,
    storybookPropsGroupID,
  );
  const labelSelector = text('label', 'Click Me!', storybookPropsGroupID);
  const isDanger = boolean('isDanger', false, storybookPropsGroupID);
  const includesIcon = boolean('includesIcon', false, storybookPropsGroupID);

  const buttonPrimaryStoryProps: ButtonPrimaryProps = {
    size: sizeSelector,
    label: labelSelector,
    isDanger,
    onPress: () => console.log("I'm clicked!"),
    width: widthSelector,
  };
  if (includesIcon) {
    const iconNameSelector = select(
      'iconName',
      IconName,
      IconName.AddSquareFilled,
      storybookPropsGroupID,
    );
    buttonPrimaryStoryProps.iconName = iconNameSelector;
  }
  return buttonPrimaryStoryProps;
};

const ButtonPrimaryStory = () => (
  <ButtonPrimary {...getButtonPrimaryStoryProps()} />
);

export default ButtonPrimaryStory;
