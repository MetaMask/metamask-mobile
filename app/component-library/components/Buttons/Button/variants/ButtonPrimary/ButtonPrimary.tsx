/* eslint-disable react/prop-types */

// Third party dependencies.
import React, { useCallback, useMemo, useState } from 'react';
import { GestureResponderEvent } from 'react-native';

// External dependencies.
import { useStyles } from '../../../../../hooks';
import Button from '../../foundation/ButtonBase';

// Internal dependencies.
import { ButtonPrimaryProps } from './ButtonPrimary.types';
import styleSheet from './ButtonPrimary.styles';

const ButtonPrimary = ({
  style,
  onPressIn,
  onPressOut,
  isDanger = false,
  ...props
}: ButtonPrimaryProps) => {
  const [pressed, setPressed] = useState(false);
  const { styles, theme } = useStyles(styleSheet, {
    style,
    isDanger,
    pressed,
  });
  const labelColor = useMemo(() => {
    const color: string = isDanger
      ? theme.colors.error.inverse
      : theme.colors.primary.inverse;
    return color;
  }, [theme, isDanger]);

  const triggerOnPressedIn = useCallback(
    (e: GestureResponderEvent) => {
      setPressed(true);
      onPressIn?.(e);
    },
    [setPressed, onPressIn],
  );

  const triggerOnPressedOut = useCallback(
    (e: GestureResponderEvent) => {
      setPressed(false);
      onPressOut?.(e);
    },
    [setPressed, onPressOut],
  );

  return (
    <Button
      style={styles.base}
      labelColor={labelColor}
      onPressIn={triggerOnPressedIn}
      onPressOut={triggerOnPressedOut}
      {...props}
    />
  );
};

export default ButtonPrimary;
