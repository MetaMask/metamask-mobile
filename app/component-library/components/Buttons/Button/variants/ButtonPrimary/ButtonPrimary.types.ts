// Internal dependencies.
import { ButtonBaseProps } from '../../foundation/ButtonBase';
import { ButtonVariants } from '../../Button.types';

/**
 * ButtonPrimary component props.
 */
export interface ButtonPrimaryProps
  extends Omit<ButtonBaseProps, 'labelColor'> {
  /**
   * Variant of Button.
   */
  variant?: ButtonVariants.Primary;
}
