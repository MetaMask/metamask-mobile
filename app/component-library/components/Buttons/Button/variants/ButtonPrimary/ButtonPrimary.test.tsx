// Third party dependencies.
import React from 'react';
import { shallow } from 'enzyme';

// External dependencies.
import { ButtonSize } from '../../Button.types';
import { IconName } from '../../../../Icon';

// Internal dependencies.
import ButtonPrimary from './ButtonPrimary';

describe('ButtonPrimary', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <ButtonPrimary
        iconName={IconName.BankFilled}
        size={ButtonSize.Md}
        label={'Click me!'}
        onPress={() => null}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
