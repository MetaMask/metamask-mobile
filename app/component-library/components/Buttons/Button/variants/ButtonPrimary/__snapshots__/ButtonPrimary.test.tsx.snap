// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonPrimary should render correctly 1`] = `
<ButtonBase
  iconName="BankFilled"
  label="Click me!"
  labelColor="#FCFCFC"
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  size="40"
  style={
    Object {
      "backgroundColor": "#0376C9",
    }
  }
/>
`;
